mariamihaelafainarea/assignment3/Assignment3

```diff
@@ -0,0 +1,192 @@
+            for(int j = 0; j < m;j++) {
+                if (dictionary.isPrefix(String.valueOf(grid[i][j]))) {
+                    //visited is used not to take more than once a cell
+                    int[][] visited = new int[n][m];
+                    visited[i][j] = 1;
+                    DFSonGrid(theList,new StringBuilder(String.valueOf(grid[i][j])),i,j,visited);
+                    visited[i][j] = 0;
+                }
+            }
+        }
+        return theList;
+    }
+    private void DFSonGrid(Set<String> theList,StringBuilder start,int x,int y,int[][] visited) {
+        //the neighbours will be vertical horizontal and diagonal.
+        //Simple DFS that also verifies if it has
+        //any reason to take a new neighbour and
+        //to expand the word with that neighbour.
+        String currentString = start.toString();
+        int[] dx = {-1,-1,-1,0,0,1,1,1};
+        int[] dy = {-1,0,1,-1,1,-1,0,1};
+        for (int i = 0; i < 8; i++) {
+            int newX = x + dx[i];
+            int newY = y + dy[i];
+            if (newX >= 0 && newX < n) {
+                if (newY >= 0 && newY < m) {
+                    if(visited[newX][newY] == 0) {
+                        visited[newX][newY] = 1;
+                        StringBuilder hopa = start.append(grid[newX][newY]);
+                        if (dictionary.isPrefix(hopa.toString())){
+                            if (dictionary.isWord(hopa.toString())) {
+                                theList.add(hopa.toString());
+                            }
+                            DFSonGrid(theList, hopa, newX, newY,visited);
+                        }
+                        //restart the state after a deep search ! 
+                        start.deleteCharAt(hopa.length()-1);
+                        visited[newX][newY] = 0;
+                    }
+                }
+            }
+        }
+        
+    }
+    class MyDictionary{
+        //A dictionary based on a trie ! 
+        //It has a better complexity than
+        //a hashmap !
+        private class TrieNode{
+            private HashMap<Character,TrieNode> children;
+            private boolean endOfWord;
+            //endOfWord is like.. here in this node
+            //a word ends even if it is not a leaf !
+            public TrieNode() {
+                children = new HashMap<>();
+                endOfWord = false;
+            }
+            
+        }
+        private final TrieNode root;
+        
+        public void insert(String word) {
+            //the idea is very simple
+            //you start from root
+            //you go deep down in the trie 
+            //with the current node
+            //if theChildren (the character that need to be added)
+            //exists you only pass it and go deep down
+            TrieNode current = root;
+            for(int i = 0; i < word.length();i++) {
+                TrieNode theChildren = current.children.get(word.charAt(i));
```

Comment:
call it theChild rather then theChildren: right now it is a little confusing because always a single TrieNode not a collection.

Reply:
done